### PR TITLE
Fantasy team endpoint

### LINKED ISSUE #17: (API) Get Fantasy Team Endpoint
- GET /fantasyteam/{id=fantasyteamId}
    * (Team details)
    * (List of players selected)
- POST /fantasyteam/

Create and get fantasyteams

### PR DESCRIPTION
**Pull request by averytribbett**
- Fantasy Team Interface 
- Fantasy Team Model
- Create Fantasy Team DTO
- Update Fantasy Team DTO
- Create Fantasy Team Service/Controller/Route/Tests
- Add NFL player to Fantasy Team Service/Controller/Route/Tests

Closes #17,  #25 , #26

### CODE REVIEW

File: rfcs/stinkball-database.md [hunk 1]
```diff
@@ -74,24 +74,21 @@ Database will store user information, NFL players, fantasy teams, and Stinkball
...
+    "qb": "number",
+    "rb1": "number",
+    "rb2": "number",
+    "wr1": "number",
+    "wr2": "number",
+    "te": "number",
+    "k": "number",
+    "dst": "number",
 }
 ```
```
Comment (Kobe16) [inline rfcs/stinkball-database.md:94]: Could you also fix the typings on the descriptions below?

File: api/src/test/fantasy-players-list.ts
Comment (Kobe16) [inline api/src/test/fantasy-players-list.ts:1]: I like that you abstracted this to a separate file

File: api/src/models/fantasy-teams.model.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
+import { Sequelize, DataTypes, Model, Optional } from 'sequelize';
+import { FantasyTeam } from '@/interfaces/fantasy-teams';
+
+export type FantasyTeamCreationAttributes = Optional<FantasyTeam, 'id' | 'team_name' | 'owner' |
+  'logo' | 'points' | 'qb' | 'wr1' | 'wr2' | 'rb1' | 'rb2' | 'te' | 'pk' | 'dst'>;
+
+
+export class FantasyTeamModel extends Model<FantasyTeam, FantasyTeamCreationAttributes> implements FantasyTeam {
```
Comment (Kobe16) [inline api/src/models/fantasy-teams.model.ts:8]: Looks good

File: api/src/dtos/fantasy-teams.dto.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  public owner: number;
+
+  @IsString()
+  @IsNotEmpty()
+  @MinLength(1)
+  public logo: string;
+
+  public points: number;
+
+  public qb: number;
```
Comment (Kobe16) [inline api/src/dtos/fantasy-teams.dto.ts:20]: For the positions, maybe you can use the Min class validator and set the minimum value to 0, because id will never be negative

File: api/src/test/fantasy-teams.test.ts [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+      expect(result.status).toEqual(201);
+      expect(result.body.message).toEqual('created');
+      expect(result.body.data.team_name).toEqual(mockFantasyTeam.team_name);
+      expect(result.body.data.owner).toEqual(mockFantasyTeam.owner);
+      expect(result.body.data.logo).toEqual(mockFantasyTeam.logo);
+      expect(result.body.data.owner).toEqual(mockFantasyTeam.owner);
+      expect(result.body.data.wr2).toEqual(null);
+    }); 
+
+    it('Create Fantasy Team Missing Auth Token (Mocking Form Submission)', async () => {
```
Comment (Kobe16) [inline api/src/test/fantasy-teams.test.ts:109]: Good idea to check for missing token

File: api/src/services/fantasy-teams.service.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import { DB } from '@database';
+import { HttpException } from '@/exceptions/httpException';
+import { FantasyTeam } from '@/interfaces/fantasy-teams';
+import { CreateFantasyTeamDto, UpdateFantasyTeamDto } from '@/dtos/fantasy-teams.dto';
+import { NFLPlayer } from '@/interfaces/nfl-players.interface';
+
+@Service()
+export class FantasyTeamService {
+  public async createFantasyTeam(fantasyTeam: CreateFantasyTeamDto): Promise<FantasyTeam> {
+    // Note we do not need to check if team already exist because teams do not need to be unique.
```
Comment (Kobe16) [inline api/src/services/fantasy-teams.service.ts:11]: good catch

File: api/src/test/fantasy-teams.test.ts [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+    it("GET fantasy team exception (Team doesn't exist)", async () => {
+      const result = await request(app.getServer())
+        .get(`${fantasyTeamRoute.path}/2`) // Check id 2 which does not exist
+        .set('Cookie', userCookie);
+
+      // Check values.
+      expect(result.status).toEqual(409);
+      expect(result.body.message).toEqual("Fantasy Team doesn't exist");
+    });
```
Comment (Kobe16) [inline api/src/test/fantasy-teams.test.ts:144]: good check

File: api/src/test/fantasy-teams.test.ts [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+      const result = await request(app.getServer())
+        .put(`${fantasyTeamRoute.path}/1`)
+        .set('Cookie', userCookie)
+        .send(addPlayer);
+
+      expect(result.status).toEqual(409);
+      expect(result.body.message).toEqual("NFL Player doesn't exist");
+    });
+
+    it('UPDATE Fantasy Team add one player (exception - Fantasy team does not exsist)', async () => {
```
Comment (Kobe16) [inline api/src/test/fantasy-teams.test.ts:217]: small typo

File: api/src/test/fantasy-teams.test.ts [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+        .put(`${fantasyTeamRoute.path}/2`)
+        .set('Cookie', userCookie)
+        .send(addPlayer);
+
+      expect(result.status).toEqual(409);
+      expect(result.body.message).toEqual("Fantasy Team doesn't exist");
+    });
+
+    it('UPDATE Fantasy Team add one player (exception - No Cookies)', async () => {
+      const addPlayer: UpdateFantasyTeamDto = {
```
Comment (Kobe16) [inline api/src/test/fantasy-teams.test.ts:232]: really great that you're going through all the edge cases

File: api/src/services/fantasy-teams.service.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    for (const [key, value] of Object.entries(newFantasyTeamData)) {
+      await this.updateIndividualRosterSpot(value, key.toString(), findFantasyTeam.id);
+    }
+    
+    const updatedFantasyTeam: FantasyTeam = await DB.FantasyTeams.findByPk(fantasyTeamId);
+    return updatedFantasyTeam;
+  }
+
+  public async updateIndividualRosterSpot(playerId: number, playerPosition: string, teamId: number): Promise<NFLPlayer> {
+    if (!playerId) return null;
```
Comment (Kobe16) [inline api/src/services/fantasy-teams.service.ts:38]: maybe you can throw an exception here if the player id is null?

File: api/src/services/fantasy-teams.service.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    // Iterate over the new fantasy team data updating the team as needed.
+    for (const [key, value] of Object.entries(newFantasyTeamData)) {
+      await this.updateIndividualRosterSpot(value, key.toString(), findFantasyTeam.id);
+    }
+    
+    const updatedFantasyTeam: FantasyTeam = await DB.FantasyTeams.findByPk(fantasyTeamId);
+    return updatedFantasyTeam;
+  }
+
+  public async updateIndividualRosterSpot(playerId: number, playerPosition: string, teamId: number): Promise<NFLPlayer> {
```
Comment (Kobe16) [inline api/src/services/fantasy-teams.service.ts:37]: I like that you abstracted this into a different function

File: api/src/services/fantasy-teams.service.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    return findFantasyTeam;
+  }
+
+  public async updateFantasyTeam(fantasyTeamId: number, newFantasyTeamData: UpdateFantasyTeamDto): Promise<FantasyTeam> {
+    const findFantasyTeam: FantasyTeam = await DB.FantasyTeams.findByPk(fantasyTeamId);
+    if (!findFantasyTeam) throw new HttpException(409, "Fantasy Team doesn't exist");
+
+    // Iterate over the new fantasy team data updating the team as needed.
+    for (const [key, value] of Object.entries(newFantasyTeamData)) {
```
Comment (Kobe16) [inline api/src/services/fantasy-teams.service.ts:29]: really cool use of for loop

**Review by Kobe16 — CHANGES_REQUESTED** [review]
Overall, this is a great PR. Pretty much everything is good from what I can see. I ran the tests and everything was able to run correctly. I made some small suggestions, but they're all minor. Good job

File: api/src/dtos/fantasy-teams.dto.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  public rb1: number;
+  public rb2: number;
+  public wr1: number;
+  public wr2: number;
+  public te: number;
+  public pk: number;
```
[the remark this answers is not in the document]
Comment (averytribbett) [reply api/src/dtos/fantasy-teams.dto.ts:26]: Yeah I realized that our external api uses PK and so it was causing issues when checking positions. We should still display K though on the front end because it makes more sense

File: api/src/services/fantasy-teams.service.ts (hunk 9 above)
Comment (averytribbett) [reply api/src/services/fantasy-teams.service.ts:38]: I think I did this so that you can update one player at a time. basically if the player is null it wont change anything with the current player

File: api/src/dtos/fantasy-teams.dto.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Kobe16) [reply api/src/dtos/fantasy-teams.dto.ts:26]: Ah ok, that's a good catch

**Review by Kobe16 — APPROVED** [review]
Looks good!
